lib/ts/recipe/session/sessionAuth.tsx

```diff
@@ -16,97 +16,82 @@
+ * It maps AuthenticationContext to SessionContext
+ */
+const SessionAuth: React.FC<Props> = ({ children, ...props }) => {
+    const [sessionContext, setSessionContext] = useState<SessionContextType>();
+    const session = Session.getInstanceOrThrow();
 
-    redirectToLogin = async (): Promise<void> => {
-        if (this.props.redirectToLogin === undefined) {
-            throw new Error(
-                "Please provide the redirectToLogin prop when using SessionAuth, or use requireAuth={false}"
-            );
+    const buildSessionContext = async (): Promise<SessionContextType> => {
+        const sessionExists = await session.doesSessionExist();
+
+        if (!sessionExists) {
+            return {
+                doesSessionExist: sessionExists,
+                jwtPayload: {},
+                userId: "",
+            };
         }
-        this.props.redirectToLogin();
+
+        const [jwtPayload, userId] = await Promise.all([session.getJWTPayloadSecurely(), session.getUserId()]);
+
+        return {
+            doesSessionExist: sessionExists,
+            jwtPayload,
+            userId,
+        };
     };
 
-    async componentDidMount(): Promise<void> {
-        const sessionExists = await doesSessionExist();
-        if (sessionExists === false) {
-            if (this.props.requireAuth !== true) {
-                this.setState((oldState) => {
-                    return {
-                        ...oldState,
-                        status: "READY",
-                        userId: "",
-                        doesSessionExist: false,
-                        jwtPayload: {},
-                    };
-                });
-            } else {
-                return await this.redirectToLogin();
-            }
-        } else {
-            const userIdPromise = getUserId();
-
-            const jwtPayloadPromise = getJWTPayloadSecurely();
-
-            const userId = await userIdPromise;
-
-            const jwtPayload = await jwtPayloadPromise;
-
-            this.setState((oldState) => {
-                return {
-                    ...oldState,
-                    status: "READY",
-                    userId,
-                    doesSessionExist: true,
-                    jwtPayload,
-                };
-            });
+    const updateSessionContext = async () => {
+        const sessionContext = await buildSessionContext();
+
+        setSessionContext(sessionContext);
+    };
+
+    const onHandleEvent = async (ctx: OnHandleEventContext) => {
+        switch (ctx.action) {
+            case "REFRESH_SESSION":
+            case "SIGN_OUT":
+            case "UNAUTHORISED":
+                return updateSessionContext();
+            default:
+                return;
         }
+    };
+
+    // Read the current state
+    useEffect(() => {
+        updateSessionContext();
+    }, []);
+
+    // Setup listener
+    useEffect(() => session.addEventListener(onHandleEvent), []);
```

Comment:
Do we need to remove eventListener when this component is unmounted?

Reply:
It will be removed, because `addEventListener` returns an unsubscribe function. `useEffect` with empty deps array will cleanup when the component is unmounted.

Reply:
Just as you advised, I will document this function better.